src/blocks/HyperKzg.sol

```diff
@@ -0,0 +1,216 @@
+        HyperKzgInput memory input,
+        uint256 a,
+        uint256[] memory D_coeffs,
+        PolyLib.UniPoly memory R_x_poly,
+        Bn256.Bn256AffinePoint memory C_P
+    ) private returns (Bn256.Bn256AffinePoint memory) {
+        Bn256.Bn256AffinePoint memory tmp = Bn256.scalarMul(input.vk_g, PolyLib.evaluate(R_x_poly, a, Bn256.R_MOD));
+        Bn256.Bn256AffinePoint memory C_K =
+            Bn256.scalarMul(input.C_Q, PolyLib.evaluate(PolyLib.UniPoly(D_coeffs), a, Bn256.R_MOD));
+        C_K = Bn256.negate(Bn256.add(tmp, C_K));
+        C_K = Bn256.add(C_P, C_K);
+        return C_K;
+    }
+
+    function pairingCheckSuccessful(HyperKzgInput memory input, Bn256.Bn256AffinePoint memory C_K, uint256 a)
+        private
+        returns (bool)
+    {
+        /*
+           let pairing_inputs: Vec<(E::G1Affine, E::G2Prepared)> = vec![
+               (C_H, vk.beta_h.into()),
+               ((C_H * (-a) - C_K).to_affine(), vk.h.into()),
+           ];
+        */
+        Pairing.G1Point memory g1_1 = Pairing.G1Point(input.C_H);
+        Pairing.G2Point memory g2_1 = input.beta_h;
+
+        Pairing.G1Point memory g1_2 =
+            Pairing.G1Point(Bn256.add(Bn256.negate(C_K), Bn256.scalarMul(input.C_H, Bn256.negateScalar(a))));
+        Pairing.G2Point memory g2_2 = input.h;
+
+        return Pairing.pairingProd2(g1_1, g2_1, g1_2, g2_2);
+    }
+
+    function verify(HyperKzgInput memory input, KeccakTranscriptLib.KeccakTranscript memory transcript)
+        public
+        returns (bool)
+    {
+        // b"c" in Rust
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x63;
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, input.pi_comms);
+        uint256 r;
+        (transcript, r) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curveBn256(), label);
+
+        uint256[] memory u = new uint256[](3);
+        u[0] = r;
+        u[1] = Bn256.negateScalar(r);
+        u[2] = mulmod(r, r, Bn256.R_MOD);
+
+        PolyLib.UniPoly memory R_x_poly = PolyLib.UniPoly(input.R_x);
+
+        // b"v" in Rust
+        label[0] = 0x76;
+        transcript = KeccakTranscriptLib.absorb(transcript, label, input.pi_evals_0, input.pi_evals_1, input.pi_evals_2);
+
+        // b"r" in Rust
+        label[0] = 0x72;
+        uint256 q;
+        (transcript, q) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curveBn256(), label);
+
+        PolyLib.UniPoly[3] memory evals =
+            [PolyLib.UniPoly(input.pi_evals_0), PolyLib.UniPoly(input.pi_evals_1), PolyLib.UniPoly(input.pi_evals_2)];
+        for (uint256 i = 0; i < evals.length; i++) {
+            if (PolyLib.evaluate(R_x_poly, u[i], Bn256.R_MOD) != PolyLib.evaluate(evals[i], q, Bn256.R_MOD)) {
+                return false;
+            }
+        }
+
+        if (!pi_polys_are_correct(input.pi_evals_0, input.pi_evals_1, input.pi_evals_2, input.point, r, input.p_of_x)) {
+            return false;
```

Comment:
I wonder if using [custom error types](https://soliditylang.org/blog/2021/04/21/custom-errors/) may not make this more maintainable, generally?

Reply:
For sure, it will make. Let's re-consider this in context of wrapping this building block into e2e contract, where whole set of errors will be defined. In assembly contract, every error should have own random 4 bytes [identifier](https://github.com/lurk-lab/solidity-verifier/blob/gas-optimizing/src/AssemblyGrumpkin.sol#L27) that allows tracing when contract is invoked via network.